internal/app/connectconformance/server_runner.go

```diff
@@ -26,8 +26,145 @@ import (
 	"connectrpc.com/conformance/internal"
 	conformancev1alpha1 "connectrpc.com/conformance/internal/gen/proto/go/connectrpc/conformance/v1alpha1"
 	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
 )
 
+// populateExpectedRequests populates the requests we expected to get back from the server
+// by examining the requests we sent.
+func populateExpectedRequests(testCase *conformancev1alpha1.TestCase) error {
+	// If an expected response was already provided, return and use that.
+	// This allows for overriding this function with explicit values in the yaml file.
+	if testCase.ExpectedResponse != nil {
+		return nil
+	}
+	if len(testCase.Request.RequestMessages) == 0 {
+		return errors.New("at least one request is required")
+	}
+
+	req := testCase.Request.RequestMessages[0]
+
+	switch testCase.Request.StreamType {
+	case conformancev1alpha1.StreamType_STREAM_TYPE_FULL_DUPLEX_BIDI_STREAM,
+		conformancev1alpha1.StreamType_STREAM_TYPE_HALF_DUPLEX_BIDI_STREAM,
+		conformancev1alpha1.StreamType_STREAM_TYPE_SERVER_STREAM:
+
+		// First, find the response definition that the client instructed the
+		// server to return
+		var def *conformancev1alpha1.StreamResponseDefinition
+		if testCase.Request.StreamType == conformancev1alpha1.StreamType_STREAM_TYPE_SERVER_STREAM {
```

Comment:
Instead of a conditional here, you could use `req.UnmarshalNew` (instead of `UnmarshalTo`) and then convert the result to an interface you can use to get the value. Something like so:
```go
concreteReq, err := req.UnmarshalNew()
if err != nil {
    return err
}
type streamResponseDefiner interface {
    GetResponseDefinition() *conformancev1alpha1.StreamResponseDefinition
}
//nolint:forcetypeassert
def = ur.(streamResponseDefiner).GetResponseDefinition()
```

And you can do similar for the types where `GetResponseDefinition()` returns a `UnaryResponseDefinition` below.

Reply:
Awesome. Updated. I added a check on the type assert but can remove if you think it's unnecessary.